module/zfs/dnode.c

```diff
@@ -676,6 +676,7 @@ dnode_reallocate(dnode_t *dn, dmu_object_type_t ot, int blocksize,
 	ASSERT(DMU_OT_IS_VALID(bonustype));
 	ASSERT3U(bonuslen, <=,
 	    DN_BONUS_SIZE(spa_maxdnodesize(dmu_objset_spa(dn->dn_objset))));
+	ASSERT3U(bonuslen, <=, DN_BONUS_SIZE(dn_slots << DNODE_SHIFT));
```

Comment:
With this ASSERT in place passing a `dn_slots == 0` no longer makes sense.  This is exactly what `dmu_object_reclaim()` does, it will need to be updated.  The following line could also be removed since it's no longer useful.

Reply:
Will do. Apologies. I meant to make this a WIP until l had the chance to write tests for it. I should have that done on Monday.